src/adc.rs

```diff
@@ -449,6 +449,244 @@ impl_adc!(ADC1: adc_unit_t_ADC_UNIT_1);
+                        &cal_config,
+                        &mut cal_handle
+                    ))
+                } {
+                    // I'd log a warning but the log crate is not available here
+                    None
+                } else {
+                    Some(cal_handle)
+                }
+            }
+            #[cfg(not(any(
+                esp32c3,
+                all(
+                    esp32c6,
+                    not(all(esp_idf_version_major = "5", esp_idf_version_minor = "0")),
+                    not(esp_idf_version_full = "5.1.0")
+                ),
+                esp32s3,
+            )))]
+            None
+        }
+
+        #[allow(unused_variables)]
+        fn get_line_calibration_handle(
+            unit_id: u8,
+            atten: adc_atten_t,
+            bitwidth: adc_bits_width_t,
+        ) -> Option<adc_cali_handle_t> {
+            #[cfg(any(esp32, esp32c2, esp32s2))]
+            {
+                // esp32 has an additional field that the exanple defalts
+                // to using fuse values for vref. Maybe we should expose
+                // this as a config option?
+                #[allow(clippy::needless_update)]
+                let cal_config = adc_cali_line_fitting_config_t {
+                    unit_id: unit_id as u32,
+                    atten,
+                    bitwidth,
+                    ..Default::default()
+                };
+                let mut cal_handle: adc_cali_handle_t = core::ptr::null_mut();
+                if let Err(_err) = unsafe {
+                    esp!(esp_idf_sys::adc_cali_create_scheme_line_fitting(
+                        &cal_config,
+                        &mut cal_handle
+                    ))
+                } {
+                    // I'd log a warning but the log crate is not available here
+                    None
+                } else {
+                    Some(cal_handle)
+                }
+            }
+            #[cfg(not(any(esp32, esp32c2, esp32s2)))]
+            None
+        }
+
+        fn pin(&mut self) -> &mut PeripheralRef<'d, T> {
+            &mut self.pin
+        }
+    }
+
+    pub struct AdcDriver<'d> {
+        handle: adc_oneshot_unit_handle_t,
+        adc: u8,
+        _ref: PhantomData<&'d ()>,
+    }
+
+    impl<'d> AdcDriver<'d> {
+        pub fn new<A: Adc>(_adc: impl Peripheral<P = A> + 'd) -> Result<Self, EspError> {
+            let config = adc_oneshot_unit_init_cfg_t {
+                unit_id: A::unit(),
+                ..Default::default()
+            };
+            let mut adc_handle: adc_oneshot_unit_handle_t = core::ptr::null_mut();
+            unsafe { esp!(adc_oneshot_new_unit(&config, &mut adc_handle))? };
+            Ok(Self {
+                handle: adc_handle,
+                adc: A::unit() as u8,
+                _ref: PhantomData,
+            })
+        }
+
+        pub fn read<T>(&mut self, channel: &mut AdcChannelDriver<'_, T>) -> Result<u16, EspError>
```

Comment:
We'll need `&self`, not `&mut self`. Also, we'll probably need `read` and `read_raw` to be implemented on the `AdcChannelDriver` side as well, delegating to these here. Why? For the case where the user instantiated `AdcChannelDriver` by passing to it `AdcDriver` by moving, rather than by reference.

Reply:
Does this open us up to any concurancy issues reading from 2 channels?

Reply:
I don't believe so. `adc_oneshot_read` [is protected by a mutex](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/adc_oneshot.html#configure-two-adc-channels), so concurrent reads on two different channels (concurrent read on a single channel is impossible, as you still pass the channel by `&mut` ref) should not be a problem.

But with that said, having a `&self` on a method does not mean that you can now use it concurrently. The structure having this method needs to also be `Sync`, or else you can't share it across several threads. And I don't believe `AdcDriver` is `Sync` by default, as it contains a raw pointer (`handle`) which makes it `!Sync` by default.

Now, with that said, I think we should make `AdcDriver` both `Sync` and `Send` (by `unsafe impl<'d, ...> Sync/Send for AdcDriver<'d, ...> {}`):
* `Send` we absolutely need, or else you can't send the `AdcDriver` to another thread post-creation.
* It also needs to be `Sync`, because otherwise `AdcChannelDriver` - which might hold a `&AdcDriver` reference - will not be `Send`, and we need `AdcChannelDriver` to implement `Send` for the same reason - so that you can send it to another thread post-creation.

As a general rule, all our drivers are `Send`, but they are `Sync` only if necessary, as using them concurrently from multiple threads is not an important use case. Again, for the 5.X oneshot driver `AdcDriver` needs to be `Sync` only because otherwise `AdcChannelDriver` cannot be `Send`; not because we want to allow concurrent access to it in the first place.